versioned/persist/store/src/main/java/org/projectnessie/versioned/persist/store/PersistVersionStore.java

```diff
@@ -0,0 +1,265 @@
+    return databaseAdapter.toHash(ref);
+  }
+
+  @Override
+  public WithHash<Ref> toRef(@Nonnull String refOfUnknownType) throws ReferenceNotFoundException {
+    try {
+      BranchName t = BranchName.of(refOfUnknownType);
+      Hash h = toHash(t);
+      return WithHash.of(h, t);
+    } catch (ReferenceNotFoundException e) {
+      TagName t = TagName.of(refOfUnknownType);
+      Hash h = toHash(t);
+      return WithHash.of(h, t);
+    }
+  }
+
+  @Override
+  public Hash commit(
+      @Nonnull BranchName branch,
+      @Nonnull Optional<Hash> expectedHead,
+      @Nonnull METADATA metadata,
+      @Nonnull List<Operation<CONTENTS>> operations)
+      throws ReferenceNotFoundException, ReferenceConflictException {
+
+    ImmutableCommitAttempt.Builder commitAttempt =
+        ImmutableCommitAttempt.builder().commitToBranch(branch).expectedHead(expectedHead);
+
+    for (Operation<CONTENTS> operation : operations) {
+      if (operation instanceof Put) {
+        Put<CONTENTS> op = (Put<CONTENTS>) operation;
+        ContentsId contentsId = ContentsId.of(storeWorker.getId(op.getValue()));
+        commitAttempt.addPuts(
+            KeyWithBytes.of(
+                op.getKey(),
+                contentsId,
+                storeWorker.getPayload(op.getValue()),
+                storeWorker.toStoreOnReferenceState(op.getValue())));
+        if (storeWorker.requiresGlobalState(op.getValue())) {
+          ByteString newState = storeWorker.toStoreGlobalState(op.getValue());
+          Optional<ByteString> expectedValue;
+          if (op.getExpectedValue() != null) {
+            if (storeWorker.getType(op.getValue()) != storeWorker.getType(op.getExpectedValue())) {
+              throw new IllegalArgumentException(
+                  String.format(
+                      "Contents-type for conditional put-operation for key '%s' for 'value' and 'expectedValue' must be the same, but are '%s' and '%s'.",
+                      op.getKey(),
+                      storeWorker.getType(op.getValue()),
+                      storeWorker.getType(op.getExpectedValue())));
+            }
+            if (!contentsId.equals(ContentsId.of(storeWorker.getId(op.getExpectedValue())))) {
+              throw new IllegalArgumentException(
+                  String.format(
+                      "Conditional put-operation key '%s' has different contents-ids.",
+                      op.getKey()));
+            }
+
+            expectedValue = Optional.of(storeWorker.toStoreGlobalState(op.getExpectedValue()));
+          } else {
+            expectedValue = Optional.empty();
+          }
+          commitAttempt.putExpectedStates(contentsId, expectedValue);
+          commitAttempt.putGlobal(contentsId, newState);
+        } else {
+          if (op.getExpectedValue() != null) {
+            // TODO shall we throw this exception? shall we validate the value somehow?
+            //  or just document the behavior?
```

Comment:
If `Expected State` means Expected _Global_ State, then I think the exception is justified in this case.